Libraries/Cordio/wsf/sources/targets/baremetal/wsf_nvm.c

```diff
@@ -402,14 +432,82 @@ bool_t WsfNvmEraseData(uint64_t id, WsfNvmCompEvent_t compCback)
 /*************************************************************************************************/
 void WsfNvmEraseDataAll(WsfNvmCompEvent_t compCback)
 {
-  for (uint32_t eraseAddr = WSF_NVM_START_ADDR; eraseAddr < wsfNvmCb.availAddr; eraseAddr += wsfNvmCb.sectorSize)
-  {
-    PalFlashEraseSector(wsfNvmCb.availAddr, eraseAddr);
-  }
-  wsfNvmCb.availAddr = WSF_NVM_START_ADDR;
-
-  if (compCback)
-  {
-    compCback(TRUE);
-  }
+    for (uint32_t eraseAddr = WSF_NVM_START_ADDR; eraseAddr < wsfNvmCb.availAddr;
+         eraseAddr += wsfNvmCb.sectorSize) {
+        PalFlashEraseSector(1, eraseAddr);
+    }
+    wsfNvmCb.availAddr = WSF_NVM_START_ADDR;
+
+    if (compCback) {
+        compCback(TRUE);
+    }
+}
+
+/*************************************************************************************************/
+/*!
+ *  \brief  Defragment NVM
+ *  \param  copyBuf       Scratch buffer to temporarily copy NVM stored data.
+ *  \param  size          Size of copyBuf in bytes.
+ *  \return  TRUE if defragment successful. FALSE otherwise
+ * 
+ *  \note   Defragment should only be called when the storage is full and a record has been invalidated.
+ *          Furthermore, copyBuf must be at least the size of WSF NVM allocated flash.
+ */
+/*************************************************************************************************/
+bool_t WsfNvmDefragment(uint8_t *copyBuf, uint32_t size)
+{
+#if WSF_ASSERT_ENABLED == 1
+    WSF_ASSERT(copyBuf && size >= wsfNvmCb.totalSize);
```

Comment:
In general I would suggest checking for `WSF_ASSERT_ENABLED` _inside_ the `WSF_ASSERT` macro.  It cleans up the code and is much easier to maintain

Reply:
@Jake-Carter Yes, with the changes provided by @EricB-ADI, the original issue can now be resolved on my end by detecting that the writing of pairing keys failed due to insufficient space, defragging, and then re-attempting to store the keys, which succeeds.

Thanks!